lib/resultuploaders/s3compatible/s3compatible.rb

```diff
@@ -0,0 +1,199 @@
+        type = 'text/html' if r_name.end_with?('.html')
+        obj = @bucket.object(remote_path)
+        obj.upload_file(l_path, content_type: type)
+        update_index_file
+        @logger.info("S3Compatible file uploaded: #{remote_path}")
+        true
+      end
+    end
+
+    def update_file_content(content, r_name)
+      handle_exceptions(__method__) do
+        remote_path = "#{@path}/#{r_name}"
+        obj = @bucket.object(remote_path)
+        obj.put(body: content)
+        @logger.info("S3Compatible file content updated: #{remote_path}")
+        true
+      end
+    end
+
+    def delete_file(r_name)
+      handle_exceptions(__method__) do
+        r_path = "#{@path}/#{r_name}"
+        obj = @bucket.object(r_path)
+        obj.delete
+        update_index_file
+        @logger.info("S3Compatible file deleted: #{r_path}")
+        true
+      end
+    end
+
+    def close; end
+
+    private
+
+    def update_index_file
+      title = @path.split('/').last
+      html_content = build_html_content(title)
+      html_content += list_objects
+      save_index_file(html_content)
+    end
+
+    def build_html_content(title)
+      <<~HTML
+        <!DOCTYPE html>
+        <html lang="en">
+        <head>
+            <meta charset="UTF-8">
+            <meta name="viewport" content="width=device-width, initial-scale=1.0">
+            <title>#{title}</title>
+            <style>
+                body {
+                    font-family: Arial, sans-serif;
+                }
+                h1 {
+                    color: #333;
+                }
+                ul {
+                    list-style-type: none;
+                    padding: 0;
+                }
+                li {
+                    margin: 5px 0;
+                }
+                a {
+                    text-decoration: none;
+                    color: #007bff;
+                }
+                a:hover {
+                    text-decoration: underline;
+                }
+            </style>
+        </head>
+        <body>
+            <h1>#{title}</h1>
+            <ul>
+      HTML
+    end
+
+    def list_objects
+      objects_html = ''
+      objects = @bucket.objects(prefix: @path)
+
+      objects.each do |object|
+        file_name = object.key.split('/').last
+        next if file_name == INDEX_FILE_NAME
+
+        file_url = generate_url(object.key)
+        objects_html += "<li><a href=\"#{file_url}\">#{file_name}</a></li>\n"
+      end
+
+      "#{objects_html}</ul>\n</body>\n</html>\n"
+    end
+
+    def save_index_file(html_content)
+      obj = @bucket.object("#{@path}/#{INDEX_FILE_NAME}")
+      obj.put(body: html_content, content_type: 'text/html')
+    end
+
+    # @note It is required that the bucket has public read permissions set.
+    # This ensures that the generated URL will be accessible without authentication.
+    def generate_url(object)
+      "https://#{@bucket_name}.#{@endpoint}/#{object}"
+    end
```

Comment:
The main reason I don’t use public_url is that it is compatible with AWS but not with Alibaba Cloud OSS. I noticed that the URL format generated by public_url is like this: https://#{@endpoint}/@bucket_name}#{object}, whereas the URL generation format for Alibaba Cloud is https://#{@bucket_name}.#{@endpoint}/#{object}. These two formats are different. The latter format that I piece together also works for AWS. What do you think?"

alibaba cloud oss detais see 
https://www.alibabacloud.com/help/en/oss/user-guide/how-to-obtain-the-url-of-a-single-object-or-the-urls-of-multiple-objects#81455fb05dw1a
![image](https://github.com/user-attachments/assets/392e190a-76a8-40e8-8c45-58dde4413bae)

Reply:
That's strange. You specify `force_path_style`, but you say this needs to return a hosted-style URL.
The only logical explanation is that you need to use the path style for uploading and the hosted style for downloading, which is awkward.

I checked the documentation and it seems Alibaba Cloud OSS *always* requires to use the hosted-style.

Overview - Object Storage Service - Alibaba Cloud Documentation Center
https://www.alibabacloud.com/help/en/oss/developer-reference/overview-24
> For security reasons, you can use only virtual hosted-style URLs instead of path-style URLs when you initiate RESTful API requests. A virtual hosted-style URL contains the bucket name.

There is also documentation that explains how to use AWS's SDK with Alibaba Cloud OSS. Unfortunately, there is no example of Ruby, but at least the example for Java has: `withPathStyleAccessEnabled(false)`.

A possible explanation is that `force_path_style` is never intended to be used with Alibaba Cloud OSS, but it is functioning as a workaround for another bug during upload. In that case, it's better to investigate that bug.

Reply:
I'm sorry, I double-checked, and you are indeed correct; Alibaba Cloud OSS does not actually require force_path_style to be set to true. I might have confused it during debugging. Thank you for the reminder